cache/http/http_test.go

```diff
@@ -2,229 +2,168 @@ package http
+	testutils "github.com/buchgr/bazel-remote/utils"
 )
 
-func TestProxyReadWorks(t *testing.T) {
-	// Test that reading a blob from a proxy works and also populates the local
-	// disk cache.
+type testServer struct {
+	srv *httptest.Server
 
-	expectedData := []byte("hello world")
-	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		w.Write(expectedData)
-	}))
-
-	cacheDir := testutils.TempDir(t)
-	defer os.RemoveAll(cacheDir)
-	diskCache := disk.New(cacheDir, 100)
-
-	baseURL, err := url.Parse(ts.URL)
-	if err != nil {
-		t.Error(err)
-	}
+	mu  sync.Mutex
+	ac  map[string][]byte
+	cas map[string][]byte
+}
 
-	proxy := New(baseURL, diskCache, &http.Client{}, testutils.NewSilentLogger(),
-		testutils.NewSilentLogger())
+func (s *testServer) handler(w http.ResponseWriter, r *http.Request) {
 
-	hashBytes := sha256.Sum256(expectedData)
-	hash := hex.EncodeToString(hashBytes[:])
-	if diskCache.Contains(cache.CAS, hash) {
-		t.Fatalf("Expected the local cache to be empty")
-	}
+	fields := strings.Split(r.URL.Path, "/")
 
-	readBytes, actualSizeBytes, err := proxy.Get(cache.CAS, hash)
-	if err != nil {
-		t.Fatalf("Failed to get the blob via the http proxy: '%v'", err)
+	kindMap := s.ac
+	if fields[1] == "ac" {
+		kindMap = s.ac
+	} else if fields[1] == "cas" {
+		kindMap = s.cas
 	}
+	hash := fields[2]
 
-	actualData, err := ioutil.ReadAll(readBytes)
-	if err != nil {
-		t.Fatal(err)
-	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
 
-	if bytes.Compare(actualData, expectedData) != 0 {
-		t.Fatalf("Expected '%v' but received '%v", actualData, expectedData)
-	}
+	switch method := r.Method; method {
+	case http.MethodGet:
+		data, ok := kindMap[hash]
+		if !ok {
+			http.Error(w, "Not found", http.StatusNotFound)
+			return
+		}
+		w.Write(data)
 
-	if actualSizeBytes != int64(len(expectedData)) {
-		t.Fatalf("Expected '%d' bytes of expected data, but received '%d'", actualSizeBytes,
-			len(expectedData))
-	}
+	case http.MethodPut:
+		data, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			http.Error(w, "failed to read body", http.StatusInternalServerError)
+		}
+		kindMap[hash] = data
 
-	if !diskCache.Contains(cache.CAS, hash) {
-		t.Fatalf("Expected the blob to be cached locally.")
+	case http.MethodHead:
+		_, ok := kindMap[hash]
+		if !ok {
+			http.Error(w, "Not found", http.StatusNotFound)
+			return
+		}
 	}
 }
 
-func TestProxyWriteWorks(t *testing.T) {
-	// Test that writing to the proxy works and also populates the local
-	// disk cache.
+func newTestServer() *testServer {
+	ts := testServer{
+		ac:  make(map[string][]byte),
+		cas: make(map[string][]byte),
+	}
+	ts.srv = httptest.NewServer(http.HandlerFunc(ts.handler))
 
-	data := []byte("hello world")
-	hashBytes := sha256.Sum256(data)
-	hash := hex.EncodeToString(hashBytes[:])
+	return &ts
+}
 
-	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		defer r.Body.Close()
+func TestEverything(t *testing.T) {
```

Comment:
This test is only testing the functionality of `newTestServer`, which is only used in this test. I think it can be removed.

Reply:
Hmm, I intended to update the tests to use this backend, but somehow the actual tests got lost in the shuffle. I'll try adding them back.

Reply:
https://github.com/buchgr/bazel-remote/pull/201